### PR TITLE
Stop sandbox JVM mode

### PR DESCRIPTION
**Pull request by markusjura**
Implementation of the `sandbox stop` command for the ConductR JVM binaries.

The `sandbox stop` command is stopping both the docker containers and the JVM processes. If no containers or processes are running then no output is produced. If either the JVM or Docker stop function cannot stop the processes or containers respectively then an error is reported and the stop function returns False.

This PR also introduces additional unit tests for the `sandbox_stop_docker` module to validate the scenarios:
- No containers running
- Cannot stop containers

The PR has been tested both with ConductR 1.1.x and ConductR 2.0.0-rc.2

### CODE REVIEW

File: conductr_cli/sandbox_common.py [hunk 1]
```diff
@@ -14,6 +10,23 @@ class ConductrComponent(Enum):
                   9006}  # ConductR bundleStreamServer
 CONDUCTR_DEV_IMAGE = 'typesafe-docker-registry-for-subscribers-only.bintray.io/conductr/conductr'
 LATEST_CONDUCTR_VERSION = '1.1.11'
+RUNNING_CONDUCTR_VERSION_FILE_PATH = '{}/RUNNING_CONDUCTR_VERSION'.format(DEFAULT_CLI_SETTINGS_DIR)
+
+
+def resolve_conductr_binrary_info(image_dir, image_version):
```
Comment (fsat) [inline conductr_cli/sandbox_common.py]: There's a spelling error - perhaps `resolve_conductr_bintray_info`?

File: conductr_cli/sandbox_stop_jvm.py [hunk 2]
```diff
@@ -6,86 +15,71 @@ def stop(args):
...
-    of the sandbox image.
-    :param run_dir: directory of where ConductR agent is running from.
-    :return: the list of the ConductR agent pids.
-    """
-    pass
-
-
-def kill_processes(pids):
+    log = logging.getLogger(__name__)
+    core_info, agent_info = sandbox_common.resolve_conductr_binrary_info(args.image_dir, args.image_version)
```
Comment (fsat) [inline conductr_cli/sandbox_stop_jvm.py]: The method should be `resolve_conductr_bintray_info`

File: conductr_cli/sandbox_run_jvm.py [hunk 3]
```diff
@@ -53,6 +52,7 @@ def run(args, features):
                      This is only relevant for Docker based sandbox since the features decides what port to expose
     :return: SandboxRunResult
     """
+    sandbox_common.LATEST_VERSION=args.image_version
```
Comment (fsat) [inline conductr_cli/sandbox_run_jvm.py]: There's no constant called `LATEST_VERSION` declared in `sandbox_common`, but there's `LATEST_CONDUCTR_VERSION`.

I don't think we should be setting a constant like this (although `LATEST_VERSION` is a variable, but by convention that's it's all uppercase, it should be treated as a constant).

I don't think `LATEST_VERSION` is used anywhere, we shouldn't need to set them.

File: conductr_cli/sandbox_common.py [hunk 4]
```diff
@@ -14,6 +10,23 @@ class ConductrComponent(Enum):
                   9006}  # ConductR bundleStreamServer
 CONDUCTR_DEV_IMAGE = 'typesafe-docker-registry-for-subscribers-only.bintray.io/conductr/conductr'
 LATEST_CONDUCTR_VERSION = '1.1.11'
```
Comment (fsat) [inline conductr_cli/sandbox_common.py]: We should move this into `sandbox_run_docker`, or perhaps remove it altogether since it's not used anywhere but in test.

I'm happy to do this in my own PR.

File: conductr_cli/sandbox_stop.py [hunk 5]
```diff
@@ -1,8 +1,16 @@
...
 
-    sandbox_stop_docker.stop(args)
+    if not hasattr(args, 'image_version'):
+        with open(RUNNING_CONDUCTR_VERSION_FILE_PATH, 'r') as file:
+            args.image_version = file.read()
+
+    if args.image_version == 1:
+        sandbox_stop_docker.stop(args)
+    elif args.image_version != 1:
+        sandbox_stop_jvm.stop(args)
```
Comment (fsat) [inline conductr_cli/sandbox_stop.py]: I think we should call stop for both docker and jvm based sandbox, since I think the user would expect both sandbox to stop if they are both running.

**Review by fsat — CHANGES_REQUESTED** [review]

File: conductr_cli/sandbox_stop.py (hunk 5 above)
Comment (markusjura) [reply conductr_cli/sandbox_stop.py]: If the `sandbox_stop_docker.stop` function is called and docker is not running then we currently throw this error:

```
Error: Encountered unexpected error.
Error: Reason: CalledProcessError Command '['docker', 'ps', '--all', '--quiet', '--filter', 'name=cond-']' returned non-zero exit status 1.
Error: Further information of the error can be found in the error log file: /Users/mj/.conductr/errors.log
```

File: conductr_cli/sandbox_run_jvm.py (hunk 3 above)
Comment (markusjura) [reply conductr_cli/sandbox_run_jvm.py]: This was an oversight of something that I've tried. Removed it.

File: conductr_cli/sandbox_stop_jvm.py (hunk 2 above)
Comment (markusjura) [reply conductr_cli/sandbox_stop_jvm.py]: Renamed it to `resolve_conductr_info` because now it does not contain only binary related information.

File: conductr_cli/sandbox_common.py (hunk 4 above)
Comment (markusjura) [reply conductr_cli/sandbox_common.py]: I'll remove it.

**Comment by markusjura** [comment]
This PR also moves the constant `LATEST_CONDUCTR_VERSION` into `conductr_cli/test/data/test_constants.py` because this constants is only used during testing.

File: conductr_cli/sandbox_run_jvm.py [hunk 6]
```diff
@@ -338,13 +308,12 @@ def extract_binary(path, binary_info):
         os.rmdir(extraction_subdir)
         return extraction_dir
 
-    core_binary_info = resolve_binary_info(ConductrComponent.CORE)
-    agent_binary_info = resolve_binary_info(ConductrComponent.AGENT)
+    core_info, agent_info = sandbox_common.resolve_conductr_info(image_dir)
 
     core_binary_path, agent_binary_path = resolve_binaries()
```
Comment (fsat) [inline conductr_cli/sandbox_run_jvm.py:313]: Just a suggestion, feel free to skip if you feel this is unnecessary.

I think it will be much more readable if `core_info` and `agent_info` is passed into `resolve_binaries`, i.e.

```
     core_binary_path, agent_binary_path = resolve_binaries(core_info, agent_info)
```

Otherwise, the `resolve_binaries` as a backwardly declared reference to `core_info` and `agent_info`. Plus `core_info` and `agent_info` are passed into the `extract_binary` in the next two lines, so it's good to be consistent.

**Review by fsat — APPROVED** [review]
